mcu/src/mcu/uart_reader.py

```diff
@@ -0,0 +1,455 @@
+"""
+
+import os
+import signal
+import sys
+import threading
+import tkinter as tk
+from tkinter import messagebox, scrolledtext, ttk
+
+import click
+import numpy as np
+import plotly.graph_objects as go
+import plotly.io as pio
+import serial
+import soundfile as sf
+from plotly.subplots import make_subplots
+from serial.tools import list_ports
+
+###################### MODIFY BELOW THIS LINE #####################
+####################################################################
+
+# Default values
+PRINT_PREFIX = "SND:HEX:"
+FREQ_SAMPLING = 10200
+DEFAULT_BAUDRATE = 115200
+VAL_MAX_ADC = 4096
+VDD = 3.3
+LOG_ENABLED = False
+AUTO_DELETE_AUDIO = False
+DEFAULT_PORT_INDEX = 1  # Change this if the default port is not the first one
+
+###################### DO NOT MODIFY BELOW THIS LINE #####################
+##########################################################################
+
+
+def generate_audio(buf, file_name):
+    buf = np.asarray(buf, dtype=np.float64)
+    buf = buf - np.mean(buf)
+    buf /= max(abs(buf))
+    sf.write("audio_files/" + file_name + ".wav", buf, FREQ_SAMPLING)
+
+
+msg_counter = 0
+
+
+def plot_signal_and_fft(signal, sampling_rate):
+    global msg_counter
+    buffer_size = len(signal)
+    times = np.linspace(0, buffer_size - 1, buffer_size) * 1 / sampling_rate
+    voltage_mV = signal * VDD / VAL_MAX_ADC * 1e3
+
+    # Calculate FFT
+    n = len(voltage_mV)
+    fft_values = np.fft.fft(voltage_mV)
+    fft_values = np.abs(fft_values)[: n // 2]  # Magnitude of the FFT
+    freqs = np.fft.fftfreq(n, 1 / sampling_rate)[: n // 2]  # Associated frequencies
+
+    # Exclude the DC component (frequency 0)
+    freqs = freqs[1:]
+    fft_values = fft_values[1:]
+
+    # Create subplot with 2 graphs (time-domain + FFT)
+    fig = make_subplots(
+        rows=2, cols=1, subplot_titles=("Time-Domain Signal", "Signal FFT")
+    )
+
+    fig.update_layout(
+        title="Real-Time Acquisition",
+        xaxis_title="Time (s)",
+        yaxis_title="Voltage (mV)",
+    )
+    fig.update_yaxes(title_text="Voltage (mV)", row=1, col=1)
+    fig.update_yaxes(title_text="FFT Magnitude", row=2, col=1)
+    fig.update_xaxes(title_text="Time (s)", row=1, col=1)
+    fig.update_xaxes(title_text="Frequency (Hz)", row=2, col=1)
+
+    # Plot time-domain signal
+    fig.add_trace(
+        go.Scatter(
+            x=times, y=voltage_mV, mode="lines", name=f"Acquisition #{msg_counter}"
+        ),
+        row=1,
+        col=1,
+    )
+    # Plot FFT without DC component
+    fig.add_trace(
+        go.Scatter(x=freqs, y=fft_values, mode="lines", name=f"FFT #{msg_counter}"),
+        row=2,
+        col=1,
+    )
+
+    # Show plots
+    pio.show(fig, renderer="browser")
+
+    # Generate audio file
+    generate_audio(signal, f"acq-{msg_counter}")
+
+    msg_counter += 1
+
+
+def print_and_save_data(data):
+    if LOG_ENABLED:
+        with open("uart_reader.log", "w") as f:
+            f.write(data)
+    print(data)
```

Comment:
In general, you should avoid using `print` in production code, especially if it's just to log information.

The `logging` module has everything you need for that, and we already have set up most of the work for that, see my main comment.